crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -318,65 +318,37 @@ use std::{cell::UnsafeCell, marker::PhantomData};
 /// ```
 /// # Safety
 ///
-/// component access of `ROQueryFetch<Self>` should be a subset of `QueryFetch<Self>`
-pub unsafe trait WorldQuery: for<'w> WorldQueryGats<'w, _State = Self::State> {
-    type ReadOnly: ReadOnlyWorldQuery<State = Self::State>;
-    type State: FetchState;
-
-    /// This function manually implements variance for the query items.
-    fn shrink<'wlong: 'wshort, 'wshort>(item: QueryItem<'wlong, Self>) -> QueryItem<'wshort, Self>;
-}
-
-/// A world query that is read only.
+/// component access of `ROQueryFetch<Self>` must be a subset of `QueryFetch<Self>`
+/// and `ROQueryFetch<Self>` must match exactly the same archetypes/tables as `QueryFetch<Self>`
 ///
-/// # Safety
-///
-/// This must only be implemented for read-only [`WorldQuery`]'s.
-pub unsafe trait ReadOnlyWorldQuery: WorldQuery {}
+/// Implementor must ensure that [`WorldQuery::update_component_access`] and
+/// [`WorldQuery::update_archetype_component_access`] exactly reflects the results of
+/// [`WorldQuery::matches_component_set`], [`WorldQuery::archetype_fetch`], and
+/// [`WorldQuery::table_fetch`].
 
-/// The [`Fetch`] of a [`WorldQuery`], which declares which data it needs access to
-pub type QueryFetch<'w, Q> = <Q as WorldQueryGats<'w>>::Fetch;
-/// The item type returned when a [`WorldQuery`] is iterated over
-pub type QueryItem<'w, Q> = <<Q as WorldQueryGats<'w>>::Fetch as Fetch<'w>>::Item;
-/// The read-only [`Fetch`] of a [`WorldQuery`], which declares which data it needs access to when accessed immutably
-pub type ROQueryFetch<'w, Q> = QueryFetch<'w, <Q as WorldQuery>::ReadOnly>;
-/// The read-only variant of the item type returned when a [`WorldQuery`] is iterated over immutably
-pub type ROQueryItem<'w, Q> = QueryItem<'w, <Q as WorldQuery>::ReadOnly>;
+pub unsafe trait WorldQuery: for<'w> WorldQueryGats<'w> {
+    type ReadOnly: ReadOnlyWorldQuery<State = Self::State>;
```

Comment:
```suggestion
    /// The read-only variant of this type, which satisfies the [`ReadOnlyWorldQuery`] trait.
    ///
    /// This defaults to `Self::State`, the correct type for queries that are already read-only.
    type ReadOnly: ReadOnlyWorldQuery<State = Self::State>;
```

Reply:
there is no defaulting going on `Trait<Assoc = Foo>` is a type equality bound requiring the assoc type `Assoc` to be equal to `Foo`. Or in this case requiring `<<Self as WorldQuery>::ReadOnly as WorldQuery>::State == <Self as WorldQuery>::State`